src/Microsoft.AspNet.Mvc.Core/Formatters/AcceptHeaderParser.cs

```diff
@@ -0,0 +1,154 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using Microsoft.AspNet.Mvc.Core;
+
+namespace Microsoft.AspNet.Mvc.Formatters
+{
+    public static class AcceptHeaderParser
+    {
+        public static IList<MediaTypeSegmentWithQuality> ParseAcceptHeader(IList<string> acceptHeaders)
+        {
+            var parsedValues = new List<MediaTypeSegmentWithQuality>();
+            ParseAcceptHeader(acceptHeaders, parsedValues);
+
+            return parsedValues;
+        }
+
+        public static void ParseAcceptHeader(IList<string> acceptHeaders, IList<MediaTypeSegmentWithQuality> parsedValues)
+        {
+            if (acceptHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(acceptHeaders));
+            }
+
+            if (parsedValues == null)
+            {
+                throw new ArgumentNullException(nameof(parsedValues));
+            }
+
+            for (var i = 0; i < acceptHeaders.Count; i++)
+            {
+                var charIndex = 0;
+                var value = acceptHeaders[i];
+
+                while (!string.IsNullOrEmpty(value) && charIndex < value.Length)
+                {
+                    MediaTypeSegmentWithQuality output;
+                    if (TryParseValue(value, ref charIndex, out output))
+                    {
+                        // The entry may not contain an actual value, like Accept: application/json, , */*
+                        if (output.MediaType.HasValue)
+                        {
+                            parsedValues.Add(output);
+                        }
+                    }
+                    else
+                    {
+                        var invalidValuesError = Resources.FormatAcceptHeaderParser_ParseAcceptHeader_InvalidValues(
+                            value.Substring(charIndex));
+
+                        throw new FormatException(invalidValuesError);
+                    }
+                }
+            }
+        }
+
+        private static bool TryParseValue(string value, ref int index, out MediaTypeSegmentWithQuality parsedValue)
+        {
+            parsedValue = default(MediaTypeSegmentWithQuality);
+
+            // The accept header may be added multiple times to the request/response message. E.g.
+            // Accept: text/xml; q=1
+            // Accept:
+            // Accept: text/plain; q=0.2
+            // In this case, do not fail parsing in case one of the values is the empty string.
+            if (string.IsNullOrEmpty(value) || (index == value.Length))
+            {
+                return true;
+            }
```

Comment:
This is why I was objecting to the factoring earlier :laughing: - I expected `true` here to mean, 'has value', but it actually means 'is valid'.